File: csharp/palindrome-products/PalindromeProducts.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class PalindromeProducts
{
    public static (int, IEnumerable<(int,int)>) Largest(int min, int max)
    {
        var palindromes = GetPalindromes(min, max);

        if (!palindromes.Any())
            throw new ArgumentException("No palindromes in the given range");
        
        return palindromes.OrderByDescending(x => x.Item1).First();
    }

    public static (int, IEnumerable<(int, int)>) Smallest(int min, int max)
    {
        var palindromes = GetPalindromes(min, max);

        if (!palindromes.Any())
            throw new ArgumentException("No palindromes in the given range");
        
        return palindromes.OrderBy(x => x.Item1).First();
    }

    private static IReadOnlyCollection<(int, IEnumerable<(int, int)>)> GetPalindromes(int min, int max) =>
        GetProducts(min, max)
            .GroupBy(x => x.Item1 * x.Item2)
            .Select(g => (g.Key, g.AsEnumerable()))
            .ToList();

    private static IEnumerable<(int, int)> GetProducts(int min, int max)
    {
        if (min > max)
            throw new ArgumentException("Invalid bounds");
        
        for (var x = min; x <= max; x++)
            for (var y = min; y <= max; y++)
                if (x <= y && IsPalindrome(x * y))
                    yield return (x, y);
    }

    private static bool IsPalindrome(int n)
    {
        var digits = n.ToString();
        return digits == string.Concat(digits.Reverse());
    }
}

File: python/palindrome-products/palindrome_products.py
"""Solution to the Palindrome Products exercise."""

from typing import Tuple, Union

Product = Tuple[Union[None, int], list[list[int]]]


def is_palindrome(number: int) -> bool:
    """Check whether the given number is palindrome.

    Args:
        number (int): the number to check

    Returns:
        bool: ``True`` if the number is a palindrome
    """
    digits = str(number)
    return digits == digits[::-1]


def find_palindrome_product(factors: list[int], products: list[int]) -> Product:
    """
    Find the first palindrome product from ``products`` and the corresponding factors.

    Args:
        factors (list[int]): factors to consider
        products (list[int]): products to consider

    Returns:
        Product: the palindrome product and the corresponding factors, if any
    """
    for p in filter(is_palindrome, products):
        factor_list = []
        for f1 in factors:
            f2, rest = divmod(p, f1)
            if rest == 0 and f1 <= f2 and f2 in factors:
                factor_list.append([f1, f2])
        
        if len(factor_list) > 0:
            return p, factor_list

    return None, []


def largest(min_factor: int, max_factor: int) -> Product:
    """Given a range of numbers, find the largest palindromes which
       are products of two numbers within that range.

    :param min_factor: int with a default value of 0
    :param max_factor: int
    :return: tuple of (palindrome, iterable).
             Iterable should contain both factors of the palindrome in an arbitrary order.
    """

    if min_factor > max_factor:
        raise ValueError("min must be <= max")

    factors = range(min_factor, max_factor + 1)
    products = range(max_factor ** 2, min_factor ** 2 - 1, -1)

    return find_palindrome_product(factors, products)


def smallest(min_factor: int, max_factor: int) -> Product:
    """Given a range of numbers, find the smallest palindromes which
    are products of two numbers within that range.

    :param min_factor: int with a default value of 0
    :param max_factor: int
    :return: tuple of (palindrome, iterable).
             Iterable should contain both factors of the palindrome in an arbitrary order.
    """

    if min_factor > max_factor:
        raise ValueError("min must be <= max")

    factors = range(min_factor, max_factor + 1)
    products = range(min_factor ** 2, max_factor ** 2 + 1)

    return find_palindrome_product(factors, products)
